specification/provider/providers.md

```diff
@@ -1,5 +1,159 @@
+
+`Providers` are implementations of the `feature provider` interface, which may wrap vendor SDKs, REST API clients, or otherwise resolve flag values from the runtime environment.
+
+##### Requirement 2.2
+
+> The `feature provider` interface **MUST** define methods to resolve flag values, with parameters `flag key` (string, required), `default value` (boolean | number | string | structure, required), `evaluation context` (optional), and `evaluation options` (optional), which returns a `flag resolution` structure.
+
+```
+// example flag resolution function
+resolveBooleanValue(flagKey, defaultValue, context, options);
+```
+
+See: [flag resolution structure](../types.md#flag-resolution), [flag value resolution](../glossary.md#flag-value-resolution)
+
+##### Condition 2.3
+
+> The implementing language type system differentiates between strings, numbers, booleans and structures.
+>
+> ##### Conditional Requirement 2.3.1
+>
+> > The `feature provider` interface **MUST** define methods for typed flag resolution, including boolean, numeric, string, and structure.
+
+```
+// example boolean flag value resolution
+FlagResolution resolveBooleanValue(string flagKey, boolean defaultValue, context: EvaluationContext, options: FlagEvaluationOptions);
+
+// example string flag value resolution
+FlagResolution resolveStringValue(string flagKey, string defaultValue, context: EvaluationContext, options: FlagEvaluationOptions);
+
+// example number flag value resolution
+FlagResolution resolveNumberValue(string flagKey, number defaultValue, context: EvaluationContext, options: FlagEvaluationOptions);
+
+// example structure flag value resolution
+FlagResolution resolveStructureValue(string flagKey, JsonObject defaultValue, context: EvaluationContext, options: FlagEvaluationOptions);
+```
+
+##### Requirement 2.4
+
+> In cases of normal execution, the `provider` **MUST** populate the `flag resolution` structure's `value` field with the resolved flag value.
+
+##### Requirement 2.4
+
+> In cases of abnormal execution, the `provider` **MUST** populate the `flag resolution` structure's `value` field with the default flag value.
+
+##### Requirement 2.5
+
+> In cases of normal execution, the `provider` **SHOULD** populate the `flag resolution` structure's `variant` field with a string identifier corresponding to the returned flag value.
+
+For example, the flag value might be `3.14159265359`, and the variant field's value might be `"pi"`.
+
+The value of the variant field might only be meaningful in the context of the flag management system associated with the provider. For example, the variant may be a UUID corresponding to the variant in the flag management system, or an index corresponding to the variant in the flag management system.
+
+##### Requirement 2.6
+
+> In cases of normal execution, the `provider` **SHOULD** populate the `flag resolution` structure's `reason` field with a string indicating the semantic reason for the returned flag value.
```

Comment:
```suggestion
> The `provider` **SHOULD** populate the `flag resolution` structure's `reason` field with a string indicating the semantic reason for the returned flag value.
```

There's a provision for `ERROR` in the reason, so they should put that in there for both cases, yes?

Reply:
Yes!